sklearn/decomposition/_fastica.py

```diff
@@ -154,6 +154,28 @@ def _cube(x, fun_args):
     return x**3, (3 * x**2).mean(axis=-1)
 
 
+@validate_params(
+    {
+        "X": ["array-like"],
+        "n_components": [Interval(Integral, 1, None, closed="left"), None],
+        "algorithm": [StrOptions({"parallel", "deflation"})],
+        "whiten": [
+            Hidden(StrOptions({"warn"})),
+            StrOptions({"arbitrary-variance", "unit-variance"}),
+            "boolean",
+        ],
+        "fun": [StrOptions({"logcosh", "exp", "cube"}), callable],
+        "fun_args": [dict, None],
+        "max_iter": [Interval(Integral, 1, None, closed="left")],
+        "tol": [Interval(Real, 0.0, None, closed="left")],
+        "w_init": ["array-like", None],
+        "whiten_solver": [StrOptions({"eigh", "svd"})],
+        "random_state": ["random_state"],
```

Comment:
All those parameters will be checked by the class `FastICA`. So no need to check them. It will be enough to verify the 3 remaining parameters.

Reply:
So what about X? Leaving out X, makes `test_function_param_validation` fail because of [this](https://github.com/scikit-learn/scikit-learn/blob/b2edcdc5e42f44b1156ea0d1a724ef10386462f6/sklearn/tests/test_public_functions.py#L34) I guess. Is it a good idea to give X a `"no_validation"`?

Reply:
It will be validated when calling `fit` or `fit_transform`, I guess.

Reply:
I assume that it will test every variable in the signature. I am wondering if we should make a partial check if possible.

Reply:
Is partial validation possible?
My two first initial solutions are either bypassing X validation with `"no_validation"` constraint (but I don't know if it is for this purpose) or giving X proper validation (`"X": ["array-like"]`).
What's your suggestion?

Reply:
OK we just looped with @jeremiedbb. Indeed, he is proposing to make a double validation.
You can just revert and add back the validation of the different parameters. Sorry for the back and forth :)

Reply:
You mean double-validate every parameter or just X?

Reply:
Every parameter indeed.

Reply:
It is unfortunate that our validation framework does not allow for partial validation. Moving forward we would need to keep the constraints in sync between the function and the estimator. For this specific PR, the `fastica` function can call `FastICA()._validate_params` to validate the common parameters and `fastica` itself can validate the remaining ones: (`X` and `compute_sources`).

@jeremiedbb What do you think of adding a `allow_partial_validation` flag to `validate_params`?

Reply:
So I guess I should wait for the partial validations as it seems that the current tests does not allow it 🤔

Reply:
> So I guess I should wait for the partial validations as it seems that the current tests does not allow it

@kianelbo To clarify, this PR did not need to be updated, because our tests does not allow for partial validation. I am okay with reverting [`6cc6c77` (#24924)](https://github.com/scikit-learn/scikit-learn/pull/24924/commits/6cc6c7777b182dfe45831aa755b71b40c805830c) and having the double validation for now.

I opened #25058 to track the discussion.